### ISSUE TITLE
Different results in SchNetPack 0.3(or 1.0) and 2.0

### ISSUE THREAD

**Issue by j-matuska** [atomistic-machine-learning/schnetpack#512]
I upgraded from SchNetPack 0.3 to SchNetPack 2.0 and to verify installation and my setup I let run the same training job (the same dataset and the same split file) on SchNetPack 2.0 as was run on SchNetPack 0.3. 
I tried to set configuration of SchNetPack 2.0 (file [train_SchNetPack2.0.txt](https://github.com/atomistic-machine-learning/schnetpack/files/11019277/train_SchNetPack2.0.txt) contain python script) as close to original SchNetPack 0.3 configuration (file [args_SchNetPack0.3.txt](https://github.com/atomistic-machine-learning/schnetpack/files/11019299/args_SchNetPack0.3.txt) contain parameters from args.json ).
However SchNetPack 2.0 calculation converged very slowly and even after 5000 epoch do not reached any meaningful results (SchNetPack 0.3 reached best model in approximately 200 epochs and stopped itself after 700). The neural network was overtrained, but loss was still one magnitude larger. The prediction of test set was inaccurate, close to random.
Did I missed something in configuration?
Is difference between these versions of SchNetPack realy so big?

**Comment by j-matuska** [comment]
Hi,
thank you for quick response. I apologize for my delay.
I am pretty sure that the error is in configuration, but I am unable to find it.

My data are structure of the molecule in xyz format (non-periodic system) and docking score(DS) of the molecule. This docking score is normalized to interval from 0 to 1.

Here is tensorboard log file from calculation using python script [train_SchNetPack2.0.txt](https://github.com/atomistic-machine-learning/schnetpack/files/11019277/train_SchNetPack2.0.txt) : 
[tensorboard_logfile_script2.0.log](https://github.com/atomistic-machine-learning/schnetpack/files/11059897/tensorboard_logfile_script2.0.log)

I did my best to set hydra and YAML config files to match this setup. Then final config file is [config.yaml.txt](https://github.com/atomistic-machine-learning/schnetpack/files/11059912/config.yaml.txt) and tensorboard log file 
[tensorboard_logfile_spktrain2.0.log](https://github.com/atomistic-machine-learning/schnetpack/files/11059923/tensorboard_logfile_spktrain2.0.log)

The csv log file from previous calculation in  SchNetPack 0.3 is 
[SchNetPack0.3.csv](https://github.com/atomistic-machine-learning/schnetpack/files/11059934/SchNetPack0.3.csv)

I hope that it is not too much mess and it will be helpful.

**Comment by NiklasGebauer** [comment]
Hi @j-matuska,

sure, no worries! Thanks for providing all the files.

It seems that using `spktrain` already helped a lot to reduce the validation loss. I am not sure what's exactly wrong in your python script, but if you want to debug it further, you could compare it to what we are doing in our [cli script](https://github.com/atomistic-machine-learning/schnetpack/blob/master/src/schnetpack/cli.py). I'd suggest to stick with `spktrain` if that's possible.

However, the model is still training a lot longer and not yet reaching the same validation loss. @Stefaanhess and me compared your configs for spk 0.3 and spk 2.0 and found two things that might be the cause for the differences:

1. spk 2.0 uses exponential moving average (ema) to obtain weights for the validation and test models per default, which was not included in in spk 0.3. Therefore, the validation loss is smoothed, which probably leads to the prolonged training (it takes longer until the learning rate is reduced). While this usually improves results, you should for now remove `ema` from `callbacks` in order to reproduce the results from spk 0.3.
2. In spk 0.3, you use 50 centers in the Gaussian expansion of distances (`num_gaussians=50`). In spk 2.0, this is set to `model.representation.radial_basis.n_rbf=20`. Please increase this number to 50.

Let us know if these changes lead to the same results as you obtained with the old SchNetPack.
To use the exact same split as in your old experiments, you can set `data.split_file=<path/to/split/file/from/spk0.3>`.
Also, the default activation functions have changed in spk 2.0 (from "shifted softplus" to "silu") but I don't think this should impair your results.

There are a few other remarks that might help you with the predictions (after you reproduced your previous results):
As far as I understand, the docking score is independent of the number of atoms in the molecule. The default parameters in SchNetPack (old and new) are designed for properties that scale with the number of atoms. For example, the default aggregation function is `globals.aggregation=sum` which is ideal e.g. for the energy. In your case, I'd try `globals.aggregation=avg` to use the average of the atom-wise predictions. Additionally, you'd have to set `is_extensive=false` in the transforms `RemoveOffsets` and `AddOffsets` (or remove these transforms, as your label is already normalized to values 0-1).

Hope this helps!
Best,
Niklas

**Comment by j-matuska** [comment]
Hi @NiklasGebauer,
it is definitely possible to stick with `spktrain`. I used python script only because I assumed, after my initial calculations, that I messed up configuration of the YAML files and I tried different approach. Even now, you have find two differences that I missed.

I modified configuration to reflect propositions (removing of ema and modification of `model.representation.radial_basis.n_rbf` ), set the `split_file` to split file used in spk 0.3(It is copied in current experimental folder) and removed transformation (`RemoveOffsets`, `AddOffsets`), because I get slightly better results without them. Final config file is [config.yaml.txt](https://github.com/atomistic-machine-learning/schnetpack/files/11080128/config.yaml.txt) and tensorboard log file 
[tensorboarg_logfile_spktrain2.0_mod.log](https://github.com/atomistic-machine-learning/schnetpack/files/11080134/tensorboarg_logfile_spktrain2.0_mod.log)
The original split file from spk 0.3 caused significant increase of the validation loss.

I experimented with addition of ema and I verified that it helps. The validation loss converge to four time smaller value.
[config.yaml.txt](https://github.com/atomistic-machine-learning/schnetpack/files/11080141/config.yaml.txt)
[tensorboard_logfile_spktrain2.0_ema.log](https://github.com/atomistic-machine-learning/schnetpack/files/11080142/tensorboard_logfile_spktrain2.0_ema.log)

The use of `globals.aggregation=sum` or `globals.aggregation=avg` was investigated at the beginning of calculation in spk 0.3. There was some dependence of DS on size and `globals.aggregation=sum` provided better results. However, there is some plan to modify DS and it have to be verified that it still holds.

Thank you for useful propositions and patience.
Best,
Jan

**Comment by NiklasGebauer** [comment]
Hi @j-matuska,

you're welcome, we're glad if we can help.
If DS depends on the number of atoms `globals.aggregation=sum` makes sense!
I checked the new logs, not sure why the error increases so much with the split from your spk 0.3 experiments.
Now, it's kinda hard to debug for us why the performance is still worse than what you observed with spk 0.3. Did you consistenly get these results with the old version or was it maybe a "lucky" run with particularly good initialization?

Maybe you can do another run (or even mulitple) with a new, random split with spk 2.0 to observe the variance. Then you could go back and use these splits in experiments with spk 0.3 and see if it consistently outperforms the new models.
Since the code is not identical, it could also be the case that different sets of hyper-parameters work better with spk 0.3 than with 2.0 and the other way around.
The log of your first run with spk 2.0 ([tensorboard_logfile_spktrain2.0.log](https://github.com/atomistic-machine-learning/schnetpack/files/11059923/tensorboard_logfile_spktrain2.0.log)) was already closer to your previous results, so I am quite optimistic that you can achieve similar scores with some hyper-parameter tuning.

Another difference between the new spk and the previous version I forgot to mention is how the training is stopped. Previously, it stopped when the minimal learning rate was reached. Now, the stopping is handled by the `early_stopping` callback. It stops after `patience=200` epochs without improvement of the validation loss of more than `min_delta=0`. This does not impair your results but the information might help you to control the length of training runs in case you feel like they take too long.

Best,
Niklas

**Comment by j-matuska** [comment]
Hi @NiklasGebauer,
I did my best to check and these results were generally better in spk 0.3. Even if I made the run with the same split but different random seed and even with different split. 

However, I did some hyperparameters tuning and the change of `batch_size` from 100 to 1 in spk 2.0 decreased the validation loss to spk 0.3 level.
To investigate further, I choose different old train dataset with approx.  50 000 molecules (much bigger than previous approx. 1000) and this converged to spk 0.3 validation loss level even with original `batch size=100`. In this case lowering `batch_size` provides only minor improvements.
It seems that there is some minimal number of steps needed to extract information from dataset and in the case of small dataset and large `batch_size` it was not possible to do before the neural network was overtrainied.

I noticed that the training is controlled by different means. Therefore, I experimented with callbacks and `early_stopping`, but I only discovered, that default values are best for spk 2.0 calculations. And that there is not possible to reproduce some behavior of  spk 0.3 (at least not easy).

Thank you for you time and guidance. It seems that I find some solutions.

Best,
 Jan
